src/persist-client/src/read.rs

```diff
@@ -1029,42 +1030,47 @@ where
         &mut self,
         as_of: Antichain<T>,
     ) -> Result<Vec<((Result<K, String>, Result<V, String>), T, D)>, Since<T>> {
-        let snap = self.snapshot(as_of).await?;
+        let batches = self.machine.snapshot(&as_of).await?;
+
+        let mut consolidator = Consolidator::new(
+            as_of.clone(),
+            self.cfg.dynamic.compaction_memory_bound_bytes(),
+        );
+
+        for batch in batches {
+            for run in batch.runs() {
+                consolidator.enqueue_run(
+                    self.machine.shard_id(),
+                    &self.blob,
+                    &self.metrics,
+                    |m| &m.snapshot,
+                    &self.machine.applier.shard_metrics,
+                    &batch.desc,
+                    run,
+                );
+            }
+        }
 
         let mut contents = Vec::new();
-        let mut last_consolidate_len = 0;
-        let mut is_consolidated = true;
-        for part in snap {
-            let (part, fetched_part) = fetch_leased_part(
-                part,
-                self.blob.as_ref(),
-                Arc::clone(&self.metrics),
-                &self.metrics.read.snapshot,
-                &self.machine.applier.shard_metrics,
-                Some(&self.reader_id),
-                self.schemas.clone(),
+
+        while let Some(iter) = consolidator.next().await {
+            contents.extend(
+                iter.filter(|(_, _, t, _)| !as_of.less_than(t))
```

Comment:
Insofar as we've had to switch from `fetch_leased_part` to the lower-level `fetch_batch_part`, yeah!

I went back and forth on this a bit... I've got a version locally that uses `FetchBatchFilter` instead of just an `as_of` in the `Consolidator`, but that had some cascading effects. (Neither variant of the filter works well for compaction, for example.) This seemed like the least invasive version to start with, but I do feel like there's probably more to do here...

Reply:
> what other sort of cascading effects?

The largest one is that we need an additional loop in the `ConsolidationPartIter`, since any given tuple might be filtered out. Inserting a third level of filtering (between the filtering that `EncodedPart` does and the filtering of zero diffs in `ConsolidatingIter`) felt like a lot. Plus a few other small things. Nothing alarming, but seemed like a lot to avoid this one line.

I can make the change, but I'm most tempted to leave a TODO and come back to it once I have the rest of the usage nailed down. As you prefer!

Reply:
So, FetchBatchFilter::Snapshot does two things: it drops things after the as_of, and also it makes sure that everything emitted [is advanced to the as_of](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/fetch.rs#L139). This latter is something that persist users are contractually allowed to rely on. If we're doing the as_of filtering here, where is the advancement happening? Also, it very much seems like we want to do the as_of advancing _before_ the zero diff filtering in ConsolidatingIter.

> Nothing alarming, but seemed like a lot to avoid this one line.

My concern isn't so much that this one line exists, so much as it feels like it's breaking an internal abstraction within persist. When you get a part from Snapshot or Listen, it's already got all the necessary ts filtering and advancement logic baked in. If we have to throw that away to use Consolidator and then manually replicate it afterward, it feels maybe too easy to get copy-pasta bugs.

Reply:
> If we're doing the as_of filtering here, where is the advancement happening?

[Here!](https://github.com/MaterializeInc/materialize/pull/20831/files#diff-085f788e243cd9a8dc3dfa0948a2701bd7f8f9739d33b1ff78fc1f9d0fbc4b04R423) This is indeed before the zero-diff filtering.

Anyways sounds like you'd prefer the fetch-batch-filtery version, so I'll roll that back in!